### DISCUSSION TITLE
ARC auto-commit & change user action from commit to PR

### DISCUSSION THREAD

**Discussion by Brilator** [nfdi4plants/arcCommander#225] [Ideas]
- Let ARC commander and ARCitect auto-commit every ARCtrl action (add study, add person, etc.) step-wise reversibility (git revert last commit)
- Let users by default work on a branch and do pull-requests to branch `main` instead of commits to "save" their work. 
- Ideally, auto-commit also every manual file change, e.g. to isa files

This would give a fine-grained "undo" function for the overall ARC (not just on user-defined commits). 

Not a fully worked out idea. Thoughts from a discussion last week @HLWeil. 
And should probably also link to ARCCtrl or ARCitect @JonasLukasczyk 

@Freymaurer @kMutagene @muehlhaus

**Comment by kMutagene** [comment]
whew, that is going to be a gigantic git tree. In my experience, the idea of maximum fine-grained commits only works in theory, and gets really tedious to work with in practice. If the tooling handles this **very** well, it is feasible. However it only helps when the commit messages are meaningful, otherwise you end up with thousands of 'update' commits. I don't think this can be done in an automated manner. `update <filename>` for example is also not a helpful commit message when you have 20 of those following each other for each edit on the file. 

I think there are 2 things here that do not fit together. Automated commit (message) creation, e.g. via a LLM, is most useful when there is actually something happening in the commit that has to be summarized to make it more clear. Fine grained commits are the exact opposite of that. Also, i think what a commit should be is context-dependent. It might not make sense to change a file without changing an accompanying other file. I think this approach in general is too opinionated to fit all use cases, and git is a flexible tool that enables people to work with it how they seem fit. Also, when not implemented very well, this might seriously disrupt workflows.

**Reply by kMutagene** [reply]
OTOH, only commiting specific commands, e.g. the set of ARCCommander commands, might be feasible and helpful, because we are able to construct meaningful commit messages from that

**Reply by Brilator** [reply]
Yeah, auto-commit for **every** file change is probably a bit much and the arc commands would probably be easy to comment and would go a long way already. The mess of adding a study via command, removing the folder and not unregistering it from the isa.investigation is just annoying. 

For manual file changes I could imagine some small reminder / warning upon close, to commit and / or sync uncommitted changes

**Reply by HLWeil** [reply]
- Yes, the idea was to use exactly the ARCCommander commands as a basis. This gives the ability to create `meaningful commit messages` simply by using the `command line input`. Details should be discussed. 
- Also, this feature is really intrusive into the user workflow, so it should be easy to `turn off`. For this, we could use the already exisiting functionality around `arc config`.
- Lastly, this could be a part of solving the https://github.com/nfdi4plants/ARCCommander/issues/196 feature request, which would be a handy, but hard to implement functionality.

**Comment by Brilator** [comment]
Different topic, similar direction. 
Just cleaning my desktop. Sorry for the spam. 

![sketch-ARC-group-fork drawio](https://github.com/nfdi4plants/ARCCommander/assets/16229676/d7d6d4a8-44ed-4469-86b7-486307078e30)
